src/Compilers/CSharp/Portable/BoundTree/DecisionTreeBuilder.cs

```diff
@@ -204,8 +212,10 @@ private DecisionTree AddByValue(DecisionTree.ByType byType, BoundConstantPattern
 
             DecisionTree forType = null;
 
-            // Find an existing decision tree for the expression's type. Since this new test
-            // should logically be last, we look for the last one we can piggy-back it onto.
+            // This new type test should logically be last. However it might be the same type as the one that is already
+            // last. In that case we can produce better code by piggy-backing our new case on to the last decision.
+            // Also, the last one might be a non-overlapping type, in which case we can piggy-back onto the second-last
+            // type test.
             for (int i = byType.TypeAndDecision.Count - 1; i >= 0 && forType == null; i--)
```

Comment:
Check `forType == null` is not necessary since there is an explicit `break;` inside the loop after `forType` is set.